Win condition

PinMeister opened the pull request:
Hopefully, everything should work now!

PinMeister committed:
Win condition and move limitations

- Player can win upon forming a X
- Player does not win if there is a strikethrough
- Added move counter
- Cannot move more than one space in any direction from current position

Things left:
- Check win condition if opponent moves a piece that allows other player to win
- Add a "checked" flag to prevent infinite recursion

PinMeister committed:
Fixed win condition

Addressed issues raised in last commit. Removed recursion.

JamalG16 commented on Game.py:
```diff
@@ -38,23 +38,125 @@ def select_players(self):
+        while True:
+            if self.player1.finished and self.player2.finished:
+                print("Game finished in a draw!")
+                exit(0)
```
pls break out of the loop instead of using exit

JamalG16 commented on Game.py:
```diff
@@ -38,23 +38,125 @@ def select_players(self):
 
     def play(self):
         self.board.draw()
+        m = 1
```
please use more descriptive variable name (please rename n as well)

Songtone commented on Game.py:
```diff
@@ -38,23 +38,125 @@ def select_players(self):
+            print(self.player1.name, 'Turn', m)
+            self.moves += self.player1.play_turn(self.moves)
+            self.win_condition(self.player1)
+            self.win_condition(self.player2)
```
Is it possible to only have the win condition check  when the player has at least 5 pieces on the board? Also, can we have only one call that checks both players instead of calling win_condition twice.

PinMeister replied:
> > I'll admit I forgot about the 5 pieces rule.
> > The reason for checking win condition twice for each player is due to this:
> > "If a player gets a regular X formed and results also in another regular X for the other player,
> > putting them both in a winning configuration, then only the player who played the last token
> > wins the game. "
> > We must account for the fact that the win condition for the player who played the last token must be checked before the other. This also applies if they move a piece that makes the other player win, which requires the other player to be checked in the same turn.
> 
> It's also worth to mention that if player 1 PLACES a piece, then there is no way player 2 can win the game and thus only player 1's win condition must be checked because checking player's 2 win condition is useless.

The app exits if player 1 wins, therefore player 2's win condition is never checked.

Songtone commented on Game.py:
```diff
@@ -38,23 +38,125 @@ def select_players(self):
-            self.win_condition()
+            self.moves += self.player2.play_turn(self.moves)
+            self.win_condition(self.player2)
+            self.win_condition(self.player1)
```
same as above

Songtone commented on Game.py:
```diff
@@ -38,23 +38,125 @@ def select_players(self):
+    def win_condition(self, player):
+
+        letter_map = self.board.get_letter_map()
+        letter_array = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L"]
```
Can we get the letters straight from the letter_map instead of hard coding it here

Songtone replied:
you can use 
for key in letter_map.keys():
   letter_array.append(key)

Songtone commented on Game.py:
```diff
@@ -38,23 +38,125 @@ def select_players(self):
+                        else:
+                            print(str(self.player2.name) + " wins!")
+                            exit(0)
+        else:
```
The code is the same as above. Can we refactor it to one code. It could check both player one after the other. Also, do we have a check where if both player get a win after a move, that the last player to make a move is the winner?

Songtone commented on Game.py:
```diff
@@ -38,23 +38,125 @@ def select_players(self):
+                        if left.get_color() == right.get_color() == TileColor.WHITE:
+                            print("Strikethrough at white X centered at " + str(iterated_piece) + "!")
+                        else:
+                            print(str(self.player2.name) + " wins!")
```
I think this should be player1

PinMeister replied:
No, strikethrough is white, so if that doesn't occur, black (player 2) wins.

PinMeister replied:
I did spot a typo elsewhere though

JamalG16 commented on Player.py:
```diff
@@ -8,14 +8,20 @@ def __init__(self, board, name, color):
 
         if len(self.played_pieces) == 0:
             choice = 1
+        elif len(self.played_pieces) == 15:
```
pls declare a variable called max_pieces or something at the class level and compare against that for readability

JamalG16 commented on Player.py:
```diff
@@ -26,8 +32,19 @@ def play_turn(self):
+            return 0
         elif choice is 2:
+            print("Moves used: " + str(moves))
+            if moves < 30:
```
same thing as the max_pieces comment, declare a class variable or something and use that for comparison for readability

PinMeister commented:
I'll admit I forgot about the 5 pieces rule.

The reason for checking win condition twice for each player is due to this: 
"If a player gets a regular X formed and results also in another regular X for the other player,
putting them both in a winning configuration, then only the player who played the last token
wins the game. "

We must account for the fact that the win condition for the player who played the last token must be checked before the other. This also applies if they move a piece that makes the other player win, which requires the other player to be checked in the same turn.

Songtone commented:
Why not have it check if they last played a piece when checking the win condition.

JamalG16 commented:
> I'll admit I forgot about the 5 pieces rule.
> 
> The reason for checking win condition twice for each player is due to this:
> "If a player gets a regular X formed and results also in another regular X for the other player,
> putting them both in a winning configuration, then only the player who played the last token
> wins the game. "
> 
> We must account for the fact that the win condition for the player who played the last token must be checked before the other. This also applies if they move a piece that makes the other player win, which requires the other player to be checked in the same turn.

It's also worth to mention that if player 1 PLACES a piece, then there is no way player 2 can win the game and thus only player 1's win condition must be checked because checking player's 2 win condition is useless.The only time you should check win condition of both players is if a piece is moved, and as you said, in the right order.

PinMeister committed:
Refactoring

- Added game over flag
- Removed duplicate code

PinMeister committed:
Variable names

For player turns

DanielPri committed:
Quality of life improvements

PinMeister committed:
Fix win condition bug

Added break for when one player wins before checking other player's win condition

DanielPri committed:
debug mode implemented

DanielPri committed:
Merge branch 'debugMode' into win-condition

PinMeister committed:
30 move bug fix

30 moves default to choice 1, if player has no more tokens available, their turn automatically passes

DanielPri reviewed (approved):
nice job fixing the bugsos

Songtone commented on Game.py:
```diff
@@ -34,27 +36,99 @@ def select_players(self):
+                        print("Strikethrough at white X centered at " + str(iterated_piece) + "!")
+                    else:
+                        print(str(self.player1.name) + " wins!")
+                        self.game_over = True
```
minor detail, you can draw the board to display to the user the board that shows the winning X.

Songtone commented on Game.py:
```diff
@@ -34,27 +36,99 @@ def select_players(self):
+                        print("Strikethrough at black X centered at " + str(iterated_piece) + "!")
+                    else:
+                        print(str(self.player2.name) + " wins!")
+                        self.game_over = True
```
same as above

Songtone reviewed (changes requested):
Other than a couple of details, it looks very good the win condition check! Good job!

PinMeister committed:
Minor adjustments

- Board is displayed upon winning
- Small changes for readability

Songtone reviewed (approved):
Good job!

JamalG16 commented on Game.py:
```diff
@@ -34,27 +36,104 @@ def select_players(self):
+                if self.game_over:
+                    break
+                else:
+                    self.win_condition(self.player2.played_pieces)
```
if player places a piece, then the number of moves is not incremented, correct? If so, then can you check that if the number of moves hasn't incremented, then do not check the win condition of player 2 please. The reason for this is that if player 1 places a piece, then there is no way player 2 can win and vice versa. Hence, checking win condition for the other player when a player places a piece is useless.

JamalG16 commented on Game.py:
```diff
@@ -34,27 +36,104 @@ def select_players(self):
+                else:
+                    self.win_condition(self.player1.played_pieces)
+            player2_turn += 1
+
```
same as above.

JamalG16 commented on Game.py:
```diff
@@ -34,27 +36,104 @@ def select_players(self):
+                        break
+                elif top_left.get_color() == top_right.get_color() == bottom_left.get_color() == bottom_right.get_color() == TileColor.BLACK:
+                    if left.get_color() == right.get_color() == TileColor.WHITE:
+                        print("Strikethrough at black X centered at " + str(iterated_piece) + "!")
```
Each player has their own played_pieces if I'm not mistaken, so what you could so is pass the tile color as a parameter in the method (win_condition(player.played_pieces, player.color)). This way you don't need to check both black and white, you could just complete the check for the color passed as param.

PinMeister replied:
This code only serves to print the correct messages for strikethroughs and who wins.

JamalG16 replied:
Yes I agree, but you're passing a specific player's pieces as a parameter. If you're passing player 1 pieces as the parameter and player 1 is white, then is there a point in having the code that checks for the black pieces?

PinMeister committed:
Other player win check

Other player's win condition is not checked if number of moves remains constant.

PinMeister committed:
Win condition adjustments

Make Jamal happy

JamalG16 commented on Game.py:
```diff
@@ -61,11 +58,11 @@ def play(self):
             self.moves += self.player1.play_turn(self.moves)
             if len(self.player1.played_pieces) >= 5 or len(self.player2.played_pieces) >= 5:
-                self.win_condition(self.player1.played_pieces)
+                self.win_condition(self.player1.name, self.player1.played_pieces, TileColor.WHITE, TileColor.BLACK)
```
looks good! only thing I would suggest is instead of hard coding the tilecolors as params, you could do player.color and for the opponent color you could add a method in tilecolor to get the opposite color and then you could pass in tilecolor.getOppositeColor(player.color).

JamalG16 replied:
Or you could even just pass in the player's color and opponent player's color

PinMeister committed:
Win con params

I can't read
